ssl/statem/extensions_clnt.c

```diff
@@ -1216,6 +1217,14 @@ EXT_RETURN tls_construct_ctos_psk(SSL_CONNECTION *s, WPACKET *pkt,
 
     msgstart = WPACKET_get_curr(pkt) - msglen;
 
+    /*
+     * difference in dtls1_set_handshake_header() vs ssl3_set_handshake_header()?
+     */
+    if (SSL_CONNECTION_IS_DTLS(s)) {
+        msgstart += DTLS1_HM_HEADER_LENGTH;
+        binderoffset -= DTLS1_HM_HEADER_LENGTH;
+    }
```

Comment:
This doesn't look correct to me. RFC8446 (TLSv1.3) says this about the binder:

> Each entry in the binders list is computed as an HMAC
>    over a transcript hash (see [Section 4.4.1](https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.1)) containing a partial
>    ClientHello up to and including the PreSharedKeyExtension.identities
>    field.

So we are doing a (partial) transcript hash. 

It later goes on to describe how the transcript hash is calculated like this:

>    Many of the cryptographic computations in TLS make use of a
>    transcript hash.  This value is computed by hashing the concatenation
>    of each included handshake message, including the handshake message
>    header carrying the handshake message type and length fields, but not
>    including record layer headers. 

So we are supposed to include the handshake message header in the transcript for TLSv1.3. RFC9147 (DTLSv1.3) says this about the transcript:

>   The transcript consists of complete TLS Handshake messages
>    (reassembled as necessary).  Note that this requires removing the
>    message_seq, fragment_offset, and fragment_length fields to create
>    the Handshake structure.

and later:

>   DTLS 1.3's handshake transcript does not include the new DTLS fields,
>    which makes it have the same format as TLS 1.3.

So, IIUC, we are supposed to manufacture a TLSv1.3 style message header to feed into the binder dropping the message_seq, fragment_offset, and fragment_length fields.

dtls1_set_handshake_header() does not fill in the header at all. I think this is mostly because at the point that we call that function we don't yet know how the message will be fragmented to fit into the MTU. This decision isn't made until much later when we hit `dtls1_do_write`.

But for this binder to work we only need the same information that was in the TLSv1.3 message header. So potentially we could make `dtls1_set_handshake_header()` additionally fill in a TLSv1.3 message header (in much the same way as `ssl3_set_handshake_header` does) so that we have that data available when we construct the binder - it would then get overwritten later when `dtls1_do_write()` gets called

Reply:
Alternatively we could manufacture the TLSv1.3 message header at this point in the code so that we're only doing it where we need to.

Reply:
> additionally fill in a TLSv1.3 message header (in much the same way as ssl3_set_handshake_header does) so that we have that data available when we construct the binder

Ah, but I suppose this approach doesn't work when reading the binder sent from the peer.